dynacred/src/user.ts

```diff
@@ -0,0 +1,122 @@
+import { CredentialStructBS } from "./credentialStructBS";
+import { CredentialTransaction } from "./credentialTransaction";
+import { ICoin } from "./genesis";
+import { KeyPair } from "./keypair";
+
+/**
+ * The User-class has references to all other needed classes.
+ * Its only additional data is the private key, which is not available in any other class.
+ * To allow for multiple users in the same db, `dbBase` is prefixed to all keys used by the user.
+ *
+ * To instantiate a user class, you should use one of the Builder.retrieveUser* methods.
+ *
+ * The user class and all classes it references are built on a simple premise:
+ * the only necessary data is the credential-ID and the private key.
+ * All other data is fetched live from the chain.
+ * A caching system makes sure that this loading is fast after a first warming up of the cache.
+ * Interacting with ByzCoin is optimized to only fetch changes of the instances.
+ */
+export class User {
+    static readonly keyPriv = "private";
+    static readonly keyCredID = "credID";
+    static readonly keyMigrate = "storage/data.json";
+    static readonly versionMigrate = 1;
+    static readonly urlNewDevice = "/register/device";
+    calypso?: Calypso;
+
+    constructor(
+        public bc: ByzCoinRPC,
+        public db: IStorage,
+        public kpp: KeyPair,
+        public dbBase: string,
+        public credStructBS: CredentialStructBS,
+        public spawnerInstanceBS: BehaviorSubject<SpawnerInstance>,
+        public coinBS: CoinBS,
+        public credSignerBS: CredentialSignerBS,
+        public addressBook: AddressBook) {
+
+        const ltsID = credStructBS.credConfig.ltsID.getValue();
+        const ltsX = credStructBS.credConfig.ltsX.getValue();
+        if (ltsID && ltsID.length === 32 && ltsX) {
+            const lts = new LongTermSecret(this.bc, ltsID, ltsX);
+            this.calypso = new Calypso(lts, credSignerBS.getValue().getBaseID(), credStructBS.credCalypso);
+        }
+    }
+
+    get kiSigner(): SignerEd25519 {
+        return this.kpp.signer();
+    }
+
+    get identityDarcSigner(): IdentityDarc {
+        return new IdentityDarc({id: this.credSignerBS.getValue().getBaseID()});
+    }
+
+    async switchKey(previous: KeyPair): Promise<KeyPair> {
+        const newKP = KeyPair.rand();
+        let dbs: DarcBS;
+        for (const db of this.credSignerBS.devices.getValue()) {
+            const rules = await db.getValue().ruleMatch(Darc.ruleSign, [previous.signer()], () => undefined);
+            if (rules.length > 0) {
+                dbs = db;
+            }
+        }
+
+        if (!dbs) {
+            throw new Error("didn't find darc with that signer");
+        }
+        await this.executeTransactions((tx) => {
+            dbs.setSignEvolve(tx, newKP.signer());
+        }, 10);
+        this.kpp = newKP;
+        await this.save();
+        return newKP;
+    }
+
+    // TODO: should this be here or in CredentialStructBS?
```

Comment:
make sense in `User`, as it's not the `CredentialStructBS` that request it but the user wanted to create a new device.

Reply:
OK, removed TODO